Add epacems crosswalk to etl

aesharpe opened the pull request:
## Context
This Draft PR is about making sure the new EPA CEMS - EIA Crosswalk file (now in the datastore not stored in the repo) gets successfully integrated into the ETL. 

I thought a lot about whether it should flow through the ETL like the rest of the data in the Datastore (i.e. have it's own extract, transform, and load modules) or get addressed separately in the `glue` module (as it is currently).

Because the crosswalk doesn't provide any "new" data, I decided to keep it lumped together with the rest of the glue. From a coding perspective, I think it's complicated when we treat data that comes from the same place (i.e. the Datastore) differently (i.e. some go through the regular ETL process and some are etl-ed in glue modules like eia_epacems.py). This is ok but it was confusing to figure out what was going on when I was attempting to build the code infrastructure. Having a clear protocol for all data would be best imo, otherwise a very well documented depiction of when/why data gets treated one way vs. another. This might not be the time or place to address this, however. 

# What this PR does:

This PR changes the `eia_epacems.py` glue file:
- Changes the name to be more specific (and all references to the old name) --> `epacems_unitid_eia_plant_crosswalk.py`
- Updates the functions inside the old `epacems_unitid_eia_plant_crosswalk.py` module to grab the raw crosswalk from the datastore rather than a csv in the repo.
- Updates the transform function.
- Adds DOIs so the Datastore script can the crosswalk data from Zenodo

zaneselvans commented:
Hey @aesharpe it looks like you may have deleted the old `eia_epacems.py` file and created an entirely new `epacems_unitid_eia_plant_crosswalk.py` file, rather than using git to rename the file, which results in the version history being broken, and it being impossible to see what all changed in the file (since git thinks the entire file is new, rather than knowing that the two files are the same entity, and comparing the differences between them).

zaneselvans commented:
Also it looks like the changes broke the ETL.

zaneselvans commented on src/pudl/glue/epacems_unitid_eia_plant_crosswalk.py:
```diff
@@ -0,0 +1,92 @@
+        "camd_plant_id": "plant_id_epa",
+        "eia_plant_name": "plant_name_eia",
+        "eia_plant_id": "plant_id_eia",
+        "eia_generator_id": "generator_id_eia",
```
Should this really be `generator_id_eia`? We just call it `generator_id` in PUDL (perhaps inadvisedly)

aesharpe replied:
I thought about that...It really should be `generator_id_eia` imo, but the rest of the data uses `generator_id`. For now it's probably smart to fix those things separately. I.e.: use `generator_id` and make an issue about changing `generator_id` to `generator_id_eia` later. It's just confusing to have an unlabeled field in a crosswalk table. Wouldn't want to assume it was epa!

TrentonBush replied:
The central challenge of the EPA crosswalk is to wrangle IDs from different sources, so I'd vote for being as explicit as possible with the naming convention, like Austen did here. I think we already have a collision in the CEMS data - PUDL calls it `plant_id_eia` but the CEMS ID fields are really from CAMD, which don't always match.

zaneselvans commented on src/pudl/glue/epacems_unitid_eia_plant_crosswalk.py:
```diff
@@ -0,0 +1,92 @@
+logger = logging.getLogger(__name__)
+
+
+def extract(ds: Datastore) -> pd.DataFrame:
```
There's no explicit need for the extract/transform steps to be broken out into their own functions. Since each of these functions are only a couple of lines, and not really meant to be reusable for any other purpose it would be fine to just have all of the operations in a single function.

aesharpe replied:
Ok I thought about doing this, but wasn't sure. Nice to have confirmation here, I'll put it all in the transform.

zaneselvans replied:
Seems like `grab_split_clean()` could also be rolled into the same single `transform()` function.

aesharpe replied:
I kind of like the idea of mimicking the extract and transform process used by the non-glue tables. I.e.: at least keep the extract and the transform function for the table if not a third function to roll them all together.

zaneselvans commented on src/pudl/glue/epacems_unitid_eia_plant_crosswalk.py:
```diff
@@ -0,0 +1,92 @@
+    """Clean raw crosswalk data, drop nans, and return split tables.
+
+    Args:
+        ds (:class:datastore.Datastore): Initialized datastore.
```
You don't have to include the type in the docstring when it's provided as a type int.

This data type is formatted incorrectly for RST. If you want to embed a link to a "role" like a `class` the name of the thing needs to be enclosed in backticks:

```rst
:class:`datastore.Datastore`
```

aesharpe replied:
oh ok so just `ds: Initialized datastore.` is good enough?

aesharpe replied:
Also I think I copied `(:class:datastore.Datastore):` from somewhere else, so it might be worth it to fix that.

TrentonBush replied:
Do we use a standard docstring format? (google, numpy, sphinx, etc) I use a docstring generator that gets close but automatically adds types and I have to manually remove them

TrentonBush commented:
> you may have deleted the old `eia_epacems.py` file and created an entirely new `epacems_unitid_eia_plant_crosswalk.py` file, rather than using git to rename the file, which results in the version history being broken

TIL! I have done this a few times in client repos

aesharpe commented:
**Still to do:** Test against crosswalk values against CEMS data

zaneselvans commented on src/pudl/etl.py:
```diff
@@ -341,11 +341,20 @@ def etl_epacems(
         glue_settings: Validated ETL parameters required by this data source.
+        ds_kwargs: Keyword arguments for instantiating a PUDL datastore, so that the ETL
+            can access the raw input data.
+        generators_entity_eia: the EIA generators entity table. Used to create subsets
```
This argument is different from what's in the call signature.

zaneselvans commented on src/pudl/etl.py:
```diff
@@ -434,8 +448,9 @@ def etl(  # noqa: C901
+    logger.info(sqlite_dfs.keys())
     if datasets.get("glue", False):
-        sqlite_dfs.update(_etl_glue(datasets["glue"]))
+        sqlite_dfs.update(_etl_glue(datasets["glue"], ds_kwargs, sqlite_dfs))
```
If this only needs the `generators_entity_eia`, I think just passing in that dataframe is better. The pattern we've had of passing around all of the dataframes to everywhere makes dependencies unclear and exposes all of the dataframes to mutation.

aesharpe replied:
That's what I was doing before, and it was breaking the tests. Because the `generators_entity_eia` table doesn't exist in the sqlite_dfs dict when the solo ferc tests are run it breaks. When you pass it the sqlite_dfs dict it doesn't break because the dict still exists. The crosswalk just won't get added because it depends on `glue_settings.eia` which indicates whether eia data has also been integrated.

zaneselvans commented on src/pudl/glue/epacems_unitid_eia_plant_crosswalk.py:
```diff
@@ -0,0 +1,119 @@
+        crosswalk_clean.generator_id.str.contains(r"^0+\d+$"), "generator_id"
+    ] = crosswalk_clean.generator_id.replace({r"^0+": ""}, regex=True)
+
+    # NOTE: still need to see whether unit_id matches up with the values in EPA well!
```
Do you mean the values in the EPA CEMS?  Which we can't enforce with a FK constraint since the CEMS data ends up in parquet, not the database.

This *is* something that we *can* check in the nightly builds though, since they'll have all the EPA CEMS and SQLite data available. We should probably organize some tests that are only run as part of the full ETL, like the validation tests are now. Maybe @bendnorman has thoughts on where we should put them or how they should be organized.

aesharpe replied:
I was thinking it would be good to see if there were any discrepancies between the CEMS data and the crosswalk. Even though there is no foreign key relationship, we still want users to be able to map the crosswalk onto the relevant dataset and that requires accurate pairing. Right above this comment I did something similar for the eia `generator_id` where I noticed that some values in the crosswalk had extra zeros in front of them. I'd want to do the same quality assurance check with CEMS.

zaneselvans replied:
This seems like a new check that we should add to the data validation tests that can get run when we've got all the processed outputs available via `tox -e nuke` or in the nightly builds.

zaneselvans commented on src/pudl/glue/epacems_unitid_eia_plant_crosswalk.py:
```diff
@@ -0,0 +1,119 @@
+        how="inner",
+    )
+
+    # There are some eia generator_id values in the crosswalk that don't match the eia
+    # generator_id values in the generators_eia860 table where the foreign keys are
+    # stored. All of them appear to have preceeding zeros. I.e.: 0010 should be 10.
+    # This makes sure to nix preceeding zeros on crosswalk generator ids that are all
+    # numeric. I.e.: 00A10 will stay 00A10 but 0010 will become 10.
+    crosswalk_clean.loc[
```
Take a look at `pudl.helpers.fix_leading_zero_gen_ids()` -- that's the function we're already using to do this kind of cleanup on the generator IDs in the EIA data itself (which contains a mix of leading zeros, or not, for the same generators... which is crazymaking). If possible it would be good to apply exactly the same fixes to both datasets so we can be sure that they match up.

zaneselvans commented on src/pudl/glue/epacems_unitid_eia_plant_crosswalk.py:
```diff
@@ -0,0 +1,119 @@
+    # tables with plant_id_eia and generator_id contain values from various years. To
+    # keep the crosswalk in alignment with the available eia data, we'll restrict it
+    # based on the generator entity table which has plant id and generator id.
+    crosswalk_clean = crosswalk_clean.merge(
+        generators_entity_eia[["plant_id_eia", "generator_id"]],
+        on=["plant_id_eia", "generator_id"],
+        how="inner",
+    )
```
This seems like a fine workaround for the fast tests, but as you note it's guaranteed to "work" even when it shouldn't, and I think we should be using all the IDs when we're doing the real full ETL or the full integration tests with all years of data.

The only two cases that we're testing and ensuring work are processing all of the years of data, or a single (most recent) year of data. We could apply this workaround conditionally, depending on what years of EIA 860 data are being processed.  If it's all available years, then we don't use the workaround, and we'll see errors that come up as a result of real mismatches between EIA 860 and the crosswalk.

In the main `etl` module you should have access to the `Eia860Settings` at some point, and can pull the years out of that, and compare them to the EIA-860 DataSource metadata, which will tell you what the full set of potentially processable years of data are.

Also, do we know if EPA is committed to updating the crosswalk year over year? As soon as we bring in the new EPA CEMS it seems like there will be new generators that aren't mapped...

aesharpe replied:
This is a good idea re: using the workaround only for the fast test.

Regarding the update frequency of the EPA crosswalk...I don't know. It seems like something that will be slow to update given the effort that went into creating it. But it's a git repo so I can open a discussion / create an issue asking this! (or see if there already is one).

zaneselvans commented on src/pudl/helpers.py:
```diff
@@ -867,39 +867,43 @@ def month_year_to_date(df):
+            logger.debug(f"Fixing leading zeros in {col_name} column")
+            df.loc[
+                df[col_name].notna()
+                & df[col_name].str.contains(number_with_leading_zeros),
```
You're performing this probably costly operation twice, instead of storing the value and reusing it. Instead I think you could do:
```py
leading_zeros = df[col_name].str.contains(number_with_leading_zeros).fillna(False)
if leading_zeros.any():
        logger.debug(f"Fixing leading zeros in {col_name} column")
        df.loc[leading_zeros, col_name] = df[col_name].str.replace(r"^0+", "", regex=True)
```

zaneselvans commented on src/pudl/metadata/fields.py:
```diff
@@ -1911,7 +1915,7 @@
     "unit_id_epa": {
         "type": "string",
-        "description": "Emissions (smokestake) unit monitored by EPA CEMS.",
+        "description": "Emissions (smokestack) unit monitored by EPA CEMS.",
```
🚭  🥩

zaneselvans commented on src/pudl/metadata/fields.py:
```diff
@@ -741,6 +741,10 @@
         "type": "string",
         "description": "Generator ID is usually numeric, but sometimes includes letters. Make sure you treat it as a string!",
     },
+    "generator_id_epa": {
```
Ugh, we should really switch to `generator_id_eia` as well. I wonder how many search-and-replaces it would take.

aesharpe replied:
I knowwww, also `boiler_id`. It's on my list of things to do once I fix everything else...

zaneselvans replied:
> once I fix everything else...

🤣🤣🤣

aesharpe commented:
**Latest changes:** 

First, I updated the crosswalk from three tables to one. The previous tables figured that there was a 1:1 relationship between `plant_id_epa` and `plant_id_eia`. All the fields are interconnected, and it makes most sense for them to be in one table. I also added in the `boiler_id` and `generator_id` fields for mapping onto more granular EIA data.

Next, I fixed some of the column names in CEMS. What we previously called `plant_id_eia` was a close but not perfect approximation of ORISPL code that the crosswalk intended to fix. I changed it to `plant_id_epa` so as not to confuse it with the official `plant_id_eia` values. There were also two unit columns in the CEMS data, one called `unitid` and one called `unit_id_epa`. `unitid` referred to the smokestack units and `unit_id_epa` referred to...well...I still don't know. So I dropped `unit_id_epa` and renamed `unitid` to `unit_id_epa`. `unit_id_epa` was also erroneously defined as the smokestack unit in our documentation so the name change meant that I could leave that description as is. Lastly, I removed the `facility_id` column because it didn't seem to refer to anything useful.

After editing the column names, I updated the metadata, ETL functions, and tests to reflect the changes.

The CEMS transform module was relying on the old `plant_id_eia` column to correct timezones in the data. While all the  values were in the right timezones, it still wasn't technically the right ID, so I pulled the crosswalk into the CEMS transform module to replace the old `plant_id_eia` column (now `plant_id_epa`) with the correct `plant_id_eia` value. I did this by filling in the long empty `harmonize_eia_epa_orispl()` function. Ultimately, I dropped the `plant_id_epa` column from CEMS in favor of the more accurate `plant_id_eia` once merged with the crosswalk. I also removed the `add_facility_id_unit_id_epa()` function because I removed those columns from the CEMS data.

aesharpe commented on src/pudl/transform/epacems.py:
```diff
@@ -208,9 +230,9 @@ def transform(raw_df: pd.DataFrame, pudl_engine: sa.engine.Engine) -> pd.DataFra
     """
     return (
         raw_df.fillna({"gross_load_mw": 0.0, "heat_content_mmbtu": 0.0})
```
Why are we doing this? To me, NA represents a lack of data where 0 represents a reading of 0. I don't think changing these values to 0 is accurate.

aesharpe replied:
I can do some more investigation, but my guess is that there was either an error with the CEMS technology or there simply wasn't a reading. It doesn't make sense that load or head content would be zero if the plant was actually functioning.

aesharpe replied:
@grgmiller do you have any thoughts on this?

grgmiller replied:
Thanks for flagging this. I would strongly recommend that NA values should definitely be left as NA, and the end user of the data can choose to fill the NA if they want. It seems like the current practice of filling NAs with zeros would contradict the "[minimize data alteration](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/data_guidelines.html#minimize-data-alteration)" principle, since you are essentially adding new data points where they do not exist in the original data.

One example of why this is important: in our hourly eGRID work, we are combining data from CEMS and EIA-923 about the hourly generation and fuel data from plants. If measured data from CEMS is available, we always prioritize it over EIA-923 data, but if CEMS is missing data, we revert to using EIA-923 data.

So in this case, if pudl is filling missing values with zero, we would be treating those zero values as measured zeros in CEMS, and overriding data from EIA-923, rather than dropping the NA records from CEMS so that we know to use EIA-923 data for that generator in that case.

zaneselvans commented on src/pudl/transform/epacems.py:
```diff
@@ -208,9 +230,9 @@ def transform(raw_df: pd.DataFrame, pudl_engine: sa.engine.Engine) -> pd.DataFra
+        .pipe(remove_leading_zeros_from_numeric_strings, "unit_id_epa")
+        .pipe(harmonize_eia_epa_orispl, pudl_engine)
         .pipe(fix_up_dates, plant_utc_offset=_load_plant_utc_offset(pudl_engine))
-        .pipe(add_facility_id_unit_id_epa)
```
Can we hunt down the true meanings of these columns (by talking to EPA through @grgmiller if need be) and document them / name them appropriately rather than removing them? We've definitely been burned by removing stuff we don't understand in the past.

Alternatively it may make more sense to integrate them into the crosswalk table (maybe they're already there under a different name?), or into the plants/generators/boilers table, since they're probably actually determined by the the plant / generator / boiler / smokestack IDs anyway and don't need to be repeated a billion times here in the data table, when they can be included in the well normalized entity tables instead.

[A bit of searching suggests](https://enviro.epa.gov/enviro/EF_METADATA_HTML.ef_metadata_page?p_column_name=FACILITY_ID&p_table_name=AA_CEMS_DETAILS&p_topic=) that `facility_id` in the CEMS context refers to the EPA Greenhouse Gas Reporting Program ID. Maybe rename it to something more legible like `ghgrp_facility_id_epa`.

Renaming the `unitid` to `unit_id_epa` and removing the column previously named `unit_id_epa` also seems like it could result in confusion, but it sounds like the metadata was previously incorrect, and the renaming makes the `unit_id_epa` actually point at the thing the metadata was describing?  "Unit ID" and "Facility ID" are so ridiculously generic. If we're swapping column names around can we give it something more descriptive and unique? e.g. `emissions_unit_id_epa`?

aesharpe replied:
> Can we hunt down the true meanings of these columns (by talking to EPA through @grgmiller if need be)

I asked Greg about them and he wasn't sure, but I'll be darned if you don't have the magic touch for finding stuff like this!! I too would prefer to keep rather than remove records.

> Alternatively it may make more sense to integrate them into the crosswalk table (maybe they're already there under a different name?),

I already compared these values to those in the crosswalk and found no meaningful overlap. I can do some grouping to see if they at least correlate with one of the other fields (plant or unit). I'll also compare them to the ghg ids you found.

> or into the plants/generators/boilers table, since they're probably actually determined by the the plant / generator / boiler / smokestack IDs anyway

Do you mean putting these EPA plant / unit labels into the EIA entity tables? Or I suppose we could fabricate another CEMS table that just had these ids in it and not all the hourly info)

> If we're swapping column names around can we give it something more descriptive and unique? e.g. emissions_unit_id_epa?

I like that idea!

aesharpe replied:
Gocha I agree. I will see what I can do!

aesharpe replied:
Made an issue for this #1772

zaneselvans commented:
As far as the naming of the crosswalk goes, what's the reporting / data umbrella that all that information is coming from, if not the EPA CEMS? Should it be `epaapmd_eia` or it looks like maybe they've rebranded AMPD to [CAMPD](https://campd.epa.gov/data).

aesharpe commented:
> As far as the naming of the crosswalk goes, what's the reporting / data umbrella that all that information is coming from, if not the EPA CEMS? Should it be `epaapmd_eia` or it looks like maybe they've rebranded AMPD to [CAMPD](https://campd.epa.gov/data) so maybe `epacampd_eia`

I made a little discussion post about it idk if you want to chat about it there or here? I think something like `epacamd_eia_crosswalk` is most accurate

grgmiller commented:
EPA suggests citing the crosswalk as:

> Huetteman, J., Tafoya, J., Johnson, T., and Schreifels, J. (2021). EPA-EIA Power Sector Data Crosswalk. Accessible at [www.epa.gov/airmarkets/power-sector-data-crosswalk](https://www.epa.gov/airmarkets/power-sector-data-crosswalk).

maybe something like `power_sector_data_crosswalk` or `epa_eia_power_sector_data_crosswalk`

grgmiller commented:
Maybe just `epa_eia_data_crosswalk` or `epa_eia_crosswalk` then?

aesharpe commented:
I think my favorite thus far is `epacamd_eia_crosswalk`

zaneselvans commented:
I don't think there's any reason to have `_data_` in the name... of course it's all data.

There are so many different EPA datasets talking about related facilities and issues, I think just saying `epa_eia_crosswalk` is a bit vague.  What's the most specific EPA program acronym that is still broadly applicable to the information contained in the crosswalk?  You're saying it's wider than the CEMS, but what all sits between EPA (as a whole agency) and CEMS (a too-narrow program)? It looks like the two main options are:

* CAMD: the [Clean Air Markets Division](https://www.epa.gov/airmarkets) (an organizational unit within EPA)
* CAMPD: [Clean Air Markets Program Data](https://campd.epa.gov/) (a particular [collection of data published by CAMD](https://campd.epa.gov/data/bulk-data-files), which includes [emissions allowances](https://gaftp.epa.gov/DMDnLoad/allowances/), [emissions themselves](https://gaftp.epa.gov/DMDnLoad/emissions/), and [compliance reporting](https://gaftp.epa.gov/DMDnLoad/compliance/))

Given our pattern of naming with the FERC and EIA data, where we concatenate the agency with the identifier of the data source (e.g. `eia`+`923` or `ferc`+`eqr`) the most analogous option here to me seems like it might be `epacampd_eia_crosswalk`.

zaneselvans commented:
I agree that just `_epa` is fine for the suffixes.

zaneselvans commented on src/pudl/glue/epacems_unitid_eia_plant_crosswalk.py:
```diff
@@ -0,0 +1,113 @@
+"""Extract, clean, and normalize the EPACAMD-EIA crosswalk.
```
Should make sure we rename this module to match whatever the newly selected name for the table is.

aesharpe replied:
Renamed. Important to note that we still need to fix the name in the scraper and archiver before we can fix the names in the pudl extractor.

zaneselvans commented on src/pudl/metadata/fields.py:
```diff
@@ -2034,16 +2033,6 @@
                 "required": True,
             }
         },
-        "plant_id_eia": {
-            "constraints": {
-                "required": True,
-            }
```
Is this being removed from the `epacems` specific resource metadata because it's no longer a required field? Isn't it one of the primary key columns in the `epacems` table? If so then it should remain required shouldn't it? Or I guess it's automatically considered required by virtue of being part of the PK.

If we're not going to fill in the NA values for `gross_load_mw` and `heat_content_mmbtu`, then those fields should also be removed from this section, since they will also no longer be required, and the NA values will violate the current schema.

aesharpe replied:
We talked about this in slack, but I'm gonna repeat it here for transparency. `plant_id_eia` has some NA values so we had to remove if from here and also demote it as a primary key in favor of the less accurate but all filled in `plant_id_epa`. The crosswalk table will now have both fields.

aesharpe commented on src/pudl/glue/epacems_unitid_eia_plant_crosswalk.py:
```diff
@@ -0,0 +1,113 @@
+    crosswalk_clean = (
+        crosswalk_clean.pipe(remove_leading_zeros_from_numeric_strings, "generator_id")
+        .pipe(remove_leading_zeros_from_numeric_strings, "unit_id_epa")
+        .dropna(subset="plant_id_eia")
```
yes, unfortunately.

zaneselvans replied:
Ah well, so it goes.

aesharpe commented on src/pudl/glue/epacems_unitid_eia_plant_crosswalk.py:
```diff
@@ -0,0 +1,113 @@
+    return {"epacamd_eia_crosswalk": crosswalk_clean}
+
+
+def crosswalk_et(
```
It felt cleaner to have a single function that executed all the code, but now that I'm looking at the structure of the other eia_etl() etc. functions in the `etl.py` module, I think it's fine to have an `extract()` function and a `transform()` function without that wrapper.

aesharpe commented on src/pudl/helpers.py:
```diff
@@ -867,39 +867,41 @@ def month_year_to_date(df):
-            generator_id=fixed_generator_id
-        )
+    if col_name in df.columns:
+        leading_zeros = df[col_name].str.contains(r"^0+\d+$")
```
In general I don't like the idea of changing `NA` data to `False`. I know it's similar, but I don't like it when there are multiple "types" of NA whether that's `np.nan`, `<NA>`, `None`, or `False`, it feels cleaner to have them all be the same/reflect the dtype of the column.

aesharpe replied:
This feels similar to what Greg was saying above about not changing data unless you have to.

zaneselvans replied:
Here `leading_zeros` is just a temporary variable which only exists within this function, that's being used as a boolean mask to select appropriate records from the dataframe for leading-zero removal, so there's no harm in tailoring it to suit that purpose (whereas above we were modifying the data that was ultimately stored in the DB). It ceases to exist as soon as the function terminates.

aesharpe replied:
> What is going on with the RTD build failure? It's timing out on some Zenodo data retrieval, but I didn't realize it even needed Zenodo data. I guess for the data source documentation?

Idk but it's working now!

aesharpe replied:
Ah you're right. Fixed!

zaneselvans commented:
What is going on with the RTD build failure? It's timing out on some Zenodo data retrieval, but I didn't realize it even needed Zenodo data. I guess for the data source documentation?

grgmiller commented:
FYI: Looks like EPA is getting ready to do a new release of the crosswalk soon, so if there are any updates or fixes that you'd like to have integrated, now might be the time to reach out: https://github.com/USEPA/camd-eia-crosswalk/pull/25#issuecomment-1190243145

aesharpe commented on src/pudl/glue/epacems_unitid_eia_plant_crosswalk.py:
```diff
@@ -0,0 +1,113 @@
+
+    # More indepth cleaning and droping rows with no plant_id_eia match.
+    crosswalk_clean = (
+        crosswalk_clean.pipe(remove_leading_zeros_from_numeric_strings, "generator_id")
```
Agreed!

aesharpe commented on src/pudl/transform/epacems.py:
```diff
@@ -191,11 +239,14 @@ def correct_gross_load_mw(df: pd.DataFrame) -> pd.DataFrame:
     # This is rare, so don't bother most of the time.
     bad = df["gross_load_mw"] > 2000
     if bad.any():
-        df.loc[bad, "gross_load_mw"] = df.loc[bad, "gross_load_mw"] / 1000
+        df.loc[bad, "gross_load_mw"] = df.gross_load_mw / 1000
     return df
```
I updated this line because I thought it was incorrect! But I'd love someone to verify.

cmgosnell replied:
i believe either are technically correct. i tend to think the previous version is more clear, but your version is more concise. the `loc` on both sides of the `=` indicates that your only trying to update/calculate a portion of the `df`. Your edit would work but you have different indexes on both sides of this equation (the column masked w/ `bad` on one side and the full column / 100 on the other. `loc` is smart enough to check the indexes and only apply the `/ 1000` version to the masked columns. I'm not sure which pattern is cleaner tbh.

grgmiller commented:
Hi @aesharpe , while you're in here updating the epacems column names, I just wanted to flag this issue for you: https://github.com/catalyst-cooperative/pudl/issues/1581

aesharpe commented:
Also ongoing convo about next crosswalk release! 
https://github.com/USEPA/camd-eia-crosswalk/issues/26

aesharpe commented on src/pudl/extract/epacems.py:
```diff
@@ -50,8 +50,8 @@
     "HEAT_INPUT": "heat_content_mmbtu",
-    "FAC_ID": "facility_id",
-    "UNIT_ID": "unit_id_epa",
+    # "FAC_ID": "facility_id",  # IDK what this is, but it isn't helpful
+    # "UNIT_ID": "unit_id_epa",  # IDK what this is, but it isn't helpful
```
Made an issue for this #1772

zaneselvans commented on src/pudl/extract/epacems.py:
```diff
@@ -36,7 +36,7 @@
-    "NOX_RATE (lbs/mmBtu)": "nox_rate_lbs_mmbtu",
+    # "NOX_RATE (lbs/mmBtu)": "nox_rate_lbs_mmbtu",
     # "NOX_RATE": "nox_rate_lbs_mmbtu",  # Not reading from CSV
     # "NOX_RATE_MEASURE_FLG": "nox_rate_measurement_code",  # Not reading from CSV
```
Don't we need all of these (including the multiple different column names) if one ever wanted to read those columns out of the raw data? It seems like leaving the mappings here, and then including them as appropriate in the list of columns to ignore would be better.  Especially knowing that a single column has different names in different years of data is useful bespoke knowledge.

aesharpe replied:
ohhh ok I was wondering why there were two ways to ignore a column. This makes a lot more sense. However, isn't the idea that rates can be calculated therefore we don't want to include them? Or maybe we should in the raw simply because they are there and there could be interesting calculation errors. Either way, we should probably keep all the rate columns or none in the raw, not just the NOX ones (unless there's a specific reason).

cmgosnell commented on src/pudl/output/epacems.py:
```diff
@@ -5,27 +5,19 @@
+def epacamd_eia_crosswalk(pudl_engine: sa.engine.Engine) -> pd.DataFrame:
+    """Pull the EPACAMD-EIA Crosswalk table."""
+    pt = pudl.output.pudltabl.get_table_meta(pudl_engine)
+    crosswalk_tbl = pt["utilities_entity_eia"]
```
is this the wrong table name?

aesharpe replied:
lol yes, luckily I fixed this a little while before.

cmgosnell commented on src/pudl/transform/epacems.py:
```diff
@@ -18,16 +18,79 @@
+    # none of the values are greater than one meaning that this drop/merge is ok. Might
+    # want to make that an official test somewhere.
+
+    crosswalk_df = pd.read_sql(
+        "epacamd_eia_crosswalk",
+        con=pudl_engine,
+        columns=["plant_id_eia", "plant_id_epa", "emissions_unit_id_epa"],
+    ).drop_duplicates()
```
Can we assume that the pudl db will have this table when this function is used? Is this managed via the etl module's sequencing of the data sets? I know there was already this coupling of cems and eia, so I assume this is not a new issue..

aesharpe replied:
This check is in the `etl_epacems()` function in the `etl` module

```
    # Verify that we have a PUDL DB with crosswalk data
    if "epacamd_eia" not in inspector.get_table_names():
        raise RuntimeError(
            "No EPACAMD-EIA Crosswalk available in the PUDL DB! Have you run the ETL? "
            f"Trying to access PUDL DB: {pudl_engine}"
        )
```

cmgosnell commented on src/pudl/transform/epacems.py:
```diff
@@ -18,16 +18,79 @@
+    # merging data together. Specifically for the convert_to_utc() function below.
+    # This creates a column based on the plant_id_eia but backfills NA with
+    # plant_id_epa so it can be used to merge on.
+    df_merged["plant_id_combined"] = df_merged.plant_id_eia.fillna(
```
is this column something that is used in the transformed and then dropped?

aesharpe replied:
Yes!

cmgosnell commented on src/pudl/transform/epacems.py:
```diff
@@ -18,16 +18,79 @@
+        df: A CEMS hourly dataframe for one year-month-state.
+
+    Returns:
+        The same data, with the ORISPL plant codes corrected to match the EIA plant IDs.
```
are the fixed or are the EIA codes just merged in? I'm not seeing replacement below, just want I think is a merge that adds the `plant_id_eia` column

aesharpe replied:
Yes, so previously the cems data had a column called `plant_id_eia`, but I changed it to `plant_id_epa` in the extract step because it was EPA's slightly inaccurate guess as to what the ORISPL code was. Most of the `plant_id_eia` values in CEMS were correct, but not all of them. The crosswalk helps fix the few cases where the EPA's ORISPL codes are bad by mapping the EPA's antiquated orispl plant code to the EIA's correct orispl code. 

This function merges the crosswalk with the cems data thus adding the official `plant_id_eia` column to CEMS. the `plant_id_combined` combines the `plant_id_eia` and `plant_id_epa` in cases where there is no `plant_id_eia` value (i.e., this plant isn't in the crosswalk yet). Because the `plant_id_epa` is almost always correct this feels reasonable!

cmgosnell commented on src/pudl/transform/epacems.py:
```diff
@@ -18,16 +18,79 @@
+    #     ~df_merged.plant_id_combined.isna().any()
+    # ), "There shouldn't be any NA vales in the combined plant id column"
+
+    assert len(df_merged) == len(df)
```
i believe you could negate the need for this assert by adding merge validation.

aesharpe replied:
nice!

cmgosnell commented on src/pudl/transform/epacems.py:
```diff
@@ -59,22 +126,28 @@ def fix_up_dates(df: pd.DataFrame, plant_utc_offset: pd.DataFrame) -> pd.DataFra
+        df["op_hour"],
+        df["op_datetime_naive"],
+        df["utc_offset"],
+        df["plant_id_combined"],
```
if you previously renamed `plant_id_eia` to `plant_id_combined` doesn't this drop also drop the original `plant_id_eia`?

aesharpe replied:
I thiiink this is based on the assumption that `plant_id_eia` previously existed in the CEMS data which is doesn't (as explained above). If not, lmk!

aesharpe replied:
Also, there is no more `plant_id_combined` anymore!

cmgosnell reviewed (changes requested):
Okay!! i have a lot of lil comments in here but mostly it looks pretty good! I'm excited to integrate this newer and better crosswalk!

One overarching comment/question... are we switching from using `cems` to `camd`? I thought that `camd` was the overarching data program of the clean air market program and `cems` is the specific dataset? like the nesting goes EPA > CAMD > CEMS. I'm not sure how the crosswalk fits in there... like does the crosswalk only pertain to CEMS or other CAMD datasets? Regardless... to me it looks like you changed some names from `cems` to `camd` and left others, which feels confusing to me.

grgmiller commented on src/pudl/transform/epacems.py:
```diff
@@ -207,10 +205,10 @@ def transform(raw_df: pd.DataFrame, pudl_engine: sa.engine.Engine) -> pd.DataFra
 
     """
     return (
-        raw_df.fillna({"gross_load_mw": 0.0, "heat_content_mmbtu": 0.0})
```
Just wanted to make sure that this was being tracked as part of https://github.com/catalyst-cooperative/pudl/issues/604

aesharpe replied:
Thanks for flagging!

zaneselvans replied:
I've connected this PR to that issue.

grgmiller commented:
Just wanted to flag that this PR will close #604

aesharpe commented:
> One overarching comment/question... are we switching from using `cems` to `camd`? I thought that `camd` was the overarching data program of the clean air market program and `cems` is the specific dataset? like the nesting goes EPA > CAMD > CEMS. I'm not sure how the crosswalk fits in there... like does the crosswalk only pertain to CEMS or other CAMD datasets? Regardless... to me it looks like you changed some names from `cems` to `camd` and left others, which feels confusing to me.

Yes, it goes EPA > CAMD > CEMS and technically the crosswalk applies to all CAMD data. It's confusing because CEMS is the only CAMD data we currently use. I decided to keep the label as CAMD rather than CEMS because a) we might want to integrate other non-cems CAMD data in the future and b) EPA calls it the CAMD crosswalk and I wanted to stay consistent with that.

aesharpe commented on src/pudl/__init__.py:
```diff
@@ -26,7 +26,7 @@
 import pudl.extract.ferc1
 import pudl.extract.ferc714
-import pudl.glue.eia_epacems
+import pudl.glue.epacamd_eia_crosswalk
```
What do you mean by the rest of PUDL?

aesharpe replied:
Maybe we should have a referendum about this lol

aesharpe replied:
Based on phone call: We are going to nix `crosswalk` from the name to keep all the glue tables the same.

aesharpe commented on src/pudl/etl.py:
```diff
@@ -371,8 +396,22 @@ def _etl_glue(glue_settings: GlueSettings) -> dict[str, pd.DataFrame]:
 
     # Add the EPA to EIA crosswalk, but only if the eia data is being processed.
     # Otherwise the foreign key references will have nothing to point at:
+    ds = Datastore(**ds_kwargs)
```
I think the reason it's outside the conditional is in case we added other glue tables that would also require the ds argument. We don't have any now though. Where do you think it should live?

cmgosnell replied:
ah okay that makes sense. its a lil weird but is more open to expansion and not at all time/memory intensive.

aesharpe commented on src/pudl/extract/epacems.py:
```diff
@@ -50,11 +50,12 @@
-    "FAC_ID": "facility_id",
-    "UNIT_ID": "unit_id_epa",
+    # "FAC_ID": "facility_id",  # unique facility id for internal EPA database management
+    # "UNIT_ID": "unit_id_what", # unique unit id for internal EPA database management
```
Some of them were commented out when I got here. Ultimately I opted to keep them all uncommented out and then add some of them to the ignore list.

zaneselvans commented on src/pudl/glue/epacamd_eia_crosswalk.py:
```diff
@@ -0,0 +1,102 @@
+    with ds.get_zipfile_resource(
+        "epacems_unitid_eia_plant_crosswalk",  # eventually change these names?
+        name="epacems_unitid_eia_plant_crosswalk.zip",  # eventually change these names?
+    ).open("camd-eia-crosswalk-master/epa_eia_crosswalk.csv") as f:
```
IIRC the scraper determines the name of the file, and I agree once we've settled on the name here, which it seems like we have, that change should be standardized across the whole chain.

aesharpe commented:
Updating the the module name in the scraper and archiver repos: 
https://github.com/catalyst-cooperative/pudl-scrapers/pull/47

zaneselvans commented:
I updated the DOIs in the datastore to pull the newly renamed archives, and (I think) updated the last of the old names in the code. I've also uploaded the new archives to gs://zenodo-cache.catalyst.coop so the CI has easy access to them.  Assuming the CI passes here, and it's attempting to load the crosswalk, then it seems like it could be merged.  All the changes from `dev` have been merged into this branch.

zaneselvans commented:
Hey @aesharpe one thing I guess this does still need is an update to the `docs/release_notes.rst` summarizing the changes.

aesharpe commented on src/pudl/metadata/resources/glue.py:
```diff
@@ -2,39 +2,20 @@
 from typing import Any
 
 RESOURCE_METADATA: dict[str, dict[str, Any]] = {
-    "assn_gen_eia_unit_epa": {
+    "epacamd_eia": {
         "schema": {
             "fields": [
```
This file is very similar to the `metadata/resources/pudl.py` file. Both have the same doc strings header: `Definitions for the glue/crosswalk tables that connect data groups.` 

The `pudl.py` file has metadata for pudl-specific tables: `plants_pudl`, `utilities_pudl`, `utility_plant_assn` and the `glue.py` file only has the `epacamd_eia` table. Should we merge them or keep them seperate?

There are some differences (pudl vs non-pudl glue), so if we decide to keep them distinct that's fine, but I'd like to clarify the doc strings so they at least say something different.

zaneselvans replied:
The stuff in `pudl.py` is limited to the manual plant and utility mapping that we do -- it's about connections between our PUDL specific IDs and other datasets, which sits pretty deep down inside the DB.

The stuff in `glue.py` is about connections between other datasets that we're bringing in.

It sounds like both docstrings are too short, and the one in `pudl.py` should be different.

aesharpe replied:
@zaneselvans

aesharpe replied:
Ok I updated the pudl one so it's different. Not going to go into too much detail rn

aesharpe commented on src/pudl/metadata/resources/glue.py:
```diff
@@ -2,39 +2,20 @@
-        "schema": {
-            "fields": [
-                "plant_id_eia",
                 "plant_id_epa",
+                "emissions_unit_id_epa",
+                "generator_id_epa",
+                "plant_id_eia",
+                "boiler_id",
+                "generator_id",
             ],
```
Right now the crosswalk table does not have any primary keys, should it?

zaneselvans replied:
What set of non-null fields can be used to uniquely identify a row in the table?

aesharpe replied:
@zaneselvans

aesharpe replied:
It's a little bit convoluted. You need everything, except technically only one of the `plant_ids`. But separating this into two tables seems too convoluted as `plant_id_eia` maps to `plant_id_eia, emission_unit_id_epa, generator_id, boiler_id`.....

zaneselvans replied:
It's pretty normal for association tables / crosswalks to have all of their columns be part of the primary key, so if you want to set all of them as part of the PK that wouldn't be crazy. So long as none of them are NA -- some databases (including SQLite) won't allow NA values in the PK.

Or if one of the Plant ID is more "real" than the other and you want to pick that one as part of the PK, but not the other plant ID (and it still uniquely identifies all the rows in the table) that would be fine too.

aesharpe replied:
Ok, sadly, there are lots of NA values in the `boiler_id` field.

zaneselvans replied:
Then an auto incrementing pseudo key 🗝️ it is!

aesharpe replied:
Do you have and recommendations for what I should call this field? Is there already a standard pseudo key name so I don't have to add something to the metadata?

aesharpe replied:
word

aesharpe commented:
> Only one line of conflict which seemed pretty clear -- the renaming of the leading zeroes fixer in the EIA-861 extractor.

How should we resolve this?

aesharpe commented on src/pudl/helpers.py:
```diff
@@ -867,39 +867,41 @@ def month_year_to_date(df):
+        if leading_zeros.any():
+            logger.debug(f"Fixing leading zeros in {col_name} column")
+            df.loc[leading_zeros, col_name] = df[col_name].str.replace(
+                r"^0+", "", regex=True
```
idk how I didn't see this comment, but I will address this!

aesharpe replied:
And the `if` thing is a very good point! I think the old one checked to see if `generator_id` was in the the df, but now that it's not column specific I can remove that line.

aesharpe replied:
NEVER MIND lol. The `if` was actually really important for how it was used in the `extract` modules for the `eia860`, `eia861`, `eia923` records. Zane found the original PR (https://github.com/catalyst-cooperative/pudl/pull/968) which shows how this function is used on raw data for these sources. 

In the future we should move this to the transform module (when we refactor all the other tables after FERC) but for now, I removed the `if` statement from the `remove_leading_zeros_from_numeric_strings` function and added it before calling the function in each of the extract modules for the aforementioned datasets.

zaneselvans commented:
Oh I resolved it by using the new name. It's all good.

aesharpe commented on src/pudl/glue/epacamd_eia_crosswalk.py:
```diff
@@ -0,0 +1,102 @@
+    # based on the generator entity table which has plant id and generator id so long
+    # as it's not using the full suite of avilable years. If it is, we don't want to
+    # restrict the crosswalk so we can get warnings and errors from any foreign key
+    # discrepancies.
```
done!

aesharpe commented on src/pudl/glue/epacamd_eia_crosswalk.py:
```diff
@@ -0,0 +1,102 @@
+            boilers_entity_eia[["plant_id_eia", "boiler_id"]],
+            on=["plant_id_eia", "boiler_id"],
+            how="inner",
+        )
```
The crosswalk doesn't have years, so the only way to restrict the crosswalk to the values that show up in the entity tables is to use plant_id and boiler_id or plant_id and generator_id. I was wondering whether there was a way to do this without merging (like with lists of tuples or something), but this was what I settled on.

cmgosnell replied:
ah got it! this makes sense. In that case, I think this is a fine method. Another way I've done this type of thing setting both df's with the same index and doing selections that way, but in this case I actually think the merge is the most straightforward.

aesharpe commented on src/pudl/glue/epacamd_eia_crosswalk.py:
```diff
@@ -0,0 +1,102 @@
+    # More indepth cleaning and droping rows with no plant_id_eia match.
+    # crosswalk_clean = crosswalk_clean.pipe(
+    #     remove_leading_zeros_from_numeric_strings, "emissions_unit_id_epa"
+    # ).dropna(subset="plant_id_eia")
```
nope, removed

cmgosnell replied:
it looks like this is still here,

zaneselvans commented on docs/release_notes.rst:
```diff
@@ -92,10 +100,18 @@ Database Schema Changes
-Date Merge Helper Function
-^^^^^^^^^^^^^^^^^^^^^^^^^^
+  started being reported, but only in the output tables. See: :pr:`1906,1911`
+* Changed and removed some columns in the :doc:`data_sources/epacems` dataset.
+  ``unitid`` was changed to ``emissions_unit_id_epa`` to clarify the type of unit it
+  represents. ``plant_id_eia`` was supplemented with values from the newly integrated
+  ``epacamd_eia`` crosswalk as not all EPA's ORISPL codes are correct. ``unit_id_epa``
+  was removed because it is a unique identifyer for ``emissions_unit_id_epa`` and not
+  otherwise useful or transferable to other datasets. ``facility_id`` was removed
+  because it is specific to EPA's internal database and does not aid in connection with
+  other data. :pr:`1692`
```
Is the change from filling zeroes to leaving NA values in this PR? Seems like a significant change we should mention.

aesharpe replied:
ohhhh yes very good catch

zaneselvans commented on notebooks/work-in-progress/play_with_cems_crosswalk.ipynb:
```diff
@@ -109,16 +108,16 @@
-      "Retrieving https://zenodo.org/api/files/4f9ac0dc-a9b4-4d2b-9e0c-c0e97a2fc7f6/epacems_unitid_eia_plant_crosswalk.zip from zenodo\n",
+      "Retrieving https://zenodo.org/api/deposit/depositions/7063255 from zenodo\n",
+      "Retrieving https://zenodo.org/api/files/afca4bd7-d94f-4af3-b1f3-abaeaec6fa68/datapackage.json from zenodo\n",
+      "Retrieving https://zenodo.org/api/files/afca4bd7-d94f-4af3-b1f3-abaeaec6fa68/epacamd_eia.zip from zenodo\n",
```
If we're done with this WIP notebook do we want to remove it?

aesharpe replied:
Sure

zaneselvans commented on src/pudl/helpers.py:
```diff
@@ -893,15 +893,14 @@ def remove_leading_zeros_from_numeric_strings(
         column.
 
     """
-    if col_name in df.columns:
```
Did you mean to remove this check? It could fail in cases where it worked before now, if it's being called on dataframes that don't have the column it's trying to fix.

aesharpe replied:
Yes -- CG had left a comment about it earlier that I forgot to address. The `if` was a relic from the function's previous form where it only accepted `generator_id`. Right now we want it to fail if you pass it a column that's not in the database rather than just move along. All instances where it's used rn are very intentional about passing it a column that should in fact be in the database.

aesharpe replied:
Resolved this by adding `if` statements prior to each time the function is called in the extract modules so that it won't break.

cmgosnell commented on src/pudl/glue/epacamd_eia_crosswalk.py:
```diff
@@ -0,0 +1,102 @@
+
+    Args:
+        epacamd_eia_crosswalk: The result of running this module's extract() function.
+        generators_entity_eia: The generators_entity_eia table.
```
It looks like this arg is still missing.

cmgosnell commented on src/pudl/etl.py:
```diff
@@ -445,8 +484,9 @@ def etl(  # noqa: C901
         sqlite_dfs.update(_etl_ferc1(datasets["ferc1"], pudl_settings))
     if datasets.get("eia", False):
         sqlite_dfs.update(_etl_eia(datasets["eia"], ds_kwargs))
+    logger.info(sqlite_dfs.keys())
```
it looks like this got left in still

cmgosnell reviewed (approved):
🎉

aesharpe commented:
This is an important after-the-fact addition: https://github.com/catalyst-cooperative/pudl/pull/1938
